util/src/main/java/io/zeebe/util/health/CriticalComponentsHealthMonitor.java

```diff
@@ -0,0 +1,141 @@
+
+  public CriticalComponentsHealthMonitor(final ActorControl actor, final Logger log) {
+    this.actor = actor;
+    this.log = log;
+  }
+
+  public void startMonitoring() {
+    actor.runAtFixedRate(HEALTH_MONITORING_PERIOD, this::updateHealth);
+  }
+
+  @Override
+  public void monitorComponent(final String componentName) {
+    actor.run(() -> componentHealth.put(componentName, HealthStatus.UNHEALTHY));
+  }
+
+  @Override
+  public void removeComponent(final String componentName) {
+    actor.run(
+        () -> {
+          monitoredComponents.remove(componentName);
+          componentHealth.remove(componentName);
+        });
+  }
+
+  @Override
+  public void registerComponent(final String componentName, final HealthMonitorable component) {
+    actor.run(
+        () -> {
+          monitoredComponents.put(componentName, component);
+          component.addFailureListener(new ComponentFailureListener(componentName));
+          componentHealth.put(componentName, component.getHealthStatus());
+          calculateHealth();
+        });
+  }
+
+  @Override
+  public HealthStatus getHealthStatus() {
+    return healthStatus;
+  }
+
+  @Override
+  public void addFailureListener(final FailureListener failureListener) {
+    actor.run(() -> this.failureListener = failureListener);
+  }
+
+  private void onComponentFailure(final String componentName) {
+    actor.run(
+        () -> {
+          log.error("{} failed, marking it as unhealthy", componentName);
+          componentHealth.put(componentName, HealthStatus.UNHEALTHY);
+          calculateHealth();
+        });
+  }
+
+  private void onComponentRecovered(final String componentName) {
+    actor.run(
+        () -> {
+          log.error("{} recovered, marking it as healthy", componentName);
+          componentHealth.put(componentName, HealthStatus.HEALTHY);
+        });
+  }
+
+  private void updateHealth() {
+    componentHealth
+        .keySet()
+        .forEach(component -> componentHealth.put(component, getHealth(component)));
+    calculateHealth();
+  }
+
+  private void calculateHealth() {
+    final var status =
+        componentHealth.containsValue(HealthStatus.UNHEALTHY)
+            ? HealthStatus.UNHEALTHY
+            : HealthStatus.HEALTHY;
+    final var previousStatus = healthStatus;
+    healthStatus = status;
+
+    if (failureListener != null && previousStatus != status) {
+      switch (status) {
+        case HEALTHY:
+          failureListener.onRecovered();
+          break;
+        case UNHEALTHY:
+          failureListener.onFailure();
+          break;
+        default:
+          break;
+      }
+    }
+  }
+
+  private HealthStatus getHealth(final String componentName) {
+    final HealthMonitorable component = monitoredComponents.get(componentName);
+    if (component != null) {
+      return component.getHealthStatus();
+    }
+    return HealthStatus.UNHEALTHY;
+  }
+
+  class ComponentFailureListener implements FailureListener {
```

Comment:
I guess here would have been a good case for us to actually have the failure listener propagate the component has argument on failure and on recovered. What do you think? Then you don't need a component listener for each component, the monitor can be a single listener.

I'm fine with this implementation for the first iteration, but we could think about it.